### PR TITLE
Migration to python-ipfabric API client

### LINKED ISSUE #71: Adopt IPfabric SDK instead of custom API client
### Proposed Functionality

Simplify code by using [IPFabric SDK](https://github.com/community-fabric/python-ipfabric)

### PR DESCRIPTION
**Pull request by pke11y**
This PR will migrate away from using the custom API client to using the `ipfabric` library for all commands.

Fixes #71

### CODE REVIEW

**Comment by pke11y** [comment]
@justinjeffery-ipf PR to migrate to v4 IP Fabric. Does not account for v5.

**Comment by justinjeffery-ipf** [comment]
Opened PR for some changes with filters.

Getting lots of errors with interfaces commands:
`Only visible to you
[10:22](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080574001300)
:warning: *An internal error occurred:
too many values to unpack (expected 2)* :warning:
Only visible to you
[10:23](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080591001400)
:warning: *An internal error occurred:
too many values to unpack (expected 2)* :warning:
Only visible to you
[10:24](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080681001500)
:warning: *An internal error occurred:
'errors'* :warning:
Only visible to you
[10:25](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080711001600)
:warning: *An internal error occurred:
interfaces() takes from 1 to 3 positional arguments but 5 were given* :warning:
Only visible to you
[10:25](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080731001700)
:warning: *An internal error occurred:
too many values to unpack (expected 2)* :warning:`

**Comment by pke11y** [comment]
> Opened PR for some changes with filters.
> 
> Getting lots of errors with interfaces commands: `Only visible to you [10:22](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080574001300) :warning: *An internal error occurred: too many values to unpack (expected 2)* :warning: Only visible to you [10:23](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080591001400) :warning: *An internal error occurred: too many values to unpack (expected 2)* :warning: Only visible to you [10:24](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080681001500) :warning: *An internal error occurred: 'errors'* :warning: Only visible to you [10:25](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080711001600) :warning: *An internal error occurred: interfaces() takes from 1 to 3 positional arguments but 5 were given* :warning: Only visible to you [10:25](https://ipfabricgroup.slack.com/archives/C02NW9MJ7CJ/p1656080731001700) :warning: *An internal error occurred: too many values to unpack (expected 2)* :warning:`

I don't think I tested interfaces, will check that.

File: nautobot_chatops_ipfabric/worker.py [hunk 1]
```diff
@@ -11,49 +11,118 @@
...
 
 BASE_CMD = "ipfabric"
 IPFABRIC_LOGO_PATH = "ipfabric/ipfabric_logo.png"
 IPFABRIC_LOGO_ALT = "IPFabric Logo"
-
+CHATOPS_IPFABRIC = "nautobot_chatops_ipfabric"
+EMPTY = "(empty)"
+
+# URLs
+INVENTORY_DEVICES_URL = "tables/inventory/devices"
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:31]: I would prefer having this in the wrapper class

File: nautobot_chatops_ipfabric/worker.py [hunk 2]
```diff
@@ -11,49 +11,118 @@
...
+INTERFACE_LOAD_URL = "tables/interfaces/load"
+INTERFACE_ERRORS_URL = "tables/interfaces/errors/bidirectional"
+INTERFACE_DROPS_URL = "tables/interfaces/drops/bidirectional"
+BGP_NEIGHBORS_URL = "tables/routing/protocols/bgp/neighbors"
+WIRELESS_SSID_URL = "tables/wireless/radio"
+WIRELESS_CLIENT_URL = "tables/wireless/clients"
+ADDRESSING_HOSTS_URL = "tables/addressing/hosts"
+
+# COLUMNS
+INVENTORY_COLUMNS = [
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:41]: I would prefer having this in the wrapper class

File: nautobot_chatops_ipfabric/worker.py [hunk 3]
```diff
@@ -198,23 +266,23 @@ def get_inventory(dispatcher, filter_key=None, filter_value=None):
                 "Device Inventory",
                 ipfabric_logo(dispatcher),
             ),
-            dispatcher.markdown_block(f"{ipfabric_api.host_url}/inventory/devices"),
+            dispatcher.markdown_block(f"{ipfabric_api.client.base_url}inventory/devices"),
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:269]: not necessary to change in this PR, but it looks like you could use the path constants from above

**Review by chadell — COMMENTED** [review]
it points in a good direction, we should validate the behavior then

File: nautobot_chatops_ipfabric/worker.py (hunk 1 above)
Comment (pke11y) [reply nautobot_chatops_ipfabric/worker.py]: agreed

File: nautobot_chatops_ipfabric/worker.py (hunk 2 above)
Comment (pke11y) [reply nautobot_chatops_ipfabric/worker.py]: agreed

File: nautobot_chatops_ipfabric/worker.py (hunk 3 above)
Comment (pke11y) [reply nautobot_chatops_ipfabric/worker.py:212]: will leave as they are for now. Need to check on the v4 UI whenever we get access. REST URL has `tables/..` as prefix.

**Comment by pke11y** [comment]
@justinjeffery-ipf should be good to go now. Fixed the interface issues.

File: nautobot_chatops_ipfabric/ipfabric_wrapper.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+from ipfabric import IPFClient
+
+# Default IP Fabric API pagination limit
+DEFAULT_PAGE_LIMIT = 100
+LAST = "$last"
+PREV = "$prev"
+LAST_LOCKED = "$lastLocked"
+
+# URLs
+INVENTORY_DEVICES_URL = "tables/inventory/devices"
```
Comment (chadell) [inline nautobot_chatops_ipfabric/ipfabric_wrapper.py:15]: my idea was to have then within the class definition. does it make sense to you?
(so you don't need to import them)

Comment (pke11y) [reply nautobot_chatops_ipfabric/ipfabric_wrapper.py:15]: Oh I like that. The importing was the reason I put them in the `worker.py`. Will update.

**Comment by justinjeffery-ipf** [comment]
Going to add some updates I thought of last night.

**Comment by pke11y** [comment]
@justinjeffery-ipf new command looks nice, thanks!
I've updated the table format to match other tables.

File: nautobot_chatops_ipfabric/worker.py [hunk 5]
```diff
@@ -832,15 +852,15 @@ def get_wireless_clients(dispatcher, ssid=None, snapshot_id=None):
...
-                client["state"],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[0]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[1]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[2]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[3]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[4]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[5]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[6]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[7]],
+                client[IpFabric.WIRELESS_CLIENT_COLUMNS[8]],
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:863]: just. a suggestion, to cleaner code, I would go with tuple comprehension like

(client[IpFabric.WIRELESS_CLIENT_COLUMNS[i]] for i in range(len(IpFabric.WIRELESS_CLIENT_COLUMNS))

File: nautobot_chatops_ipfabric/worker.py [hunk 6]
```diff
@@ -881,24 +910,24 @@ def find_host(dispatcher, filter_key=None, filter_value=None):
...
-                host.get("vrf") or "(empty)",
-                host.get("dnsName") or "(empty)",
-                host.get("siteName") or "(empty)",
-                host.get("edges") or "(empty)",
-                host.get("gateways") or "(empty)",
-                host.get("accessPoints") or "(empty)",
-                host.get("mac") or "(empty)",
-                host.get("vendor") or "(empty)",
-                host.get("vlan") or "(empty)",
+                host.get(IpFabric.ADDRESSING_HOSTS_COLUMNS[0]) or IpFabric.EMPTY,
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:921]: (host.get(IpFabric.ADDRESSING_HOSTS_COLUMNS[1], IpFabric.EMPTY) for i in range(len(IpFabric.ADDRESSING_HOSTS_COLUMNS))

File: nautobot_chatops_ipfabric/worker.py [hunk 7]
```diff
@@ -761,12 +767,12 @@ def get_wireless_ssids(dispatcher, ssid=None, snapshot_id=None):
         ],
         [
             (
-                ssid["wlanSsid"],
-                ssid["siteName"],
-                ssid["apName"],
-                ssid["radioDscr"],
-                ssid["radioStatus"],
-                ssid["clientCount"],
+                ssid[IpFabric.WIRELESS_SSID_COLUMNS[0]],
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:770]:
```py
(ssid[IpFabric.WIRELESS_SSID_COLUMNS[0]] for i in range(len(IpFabric.WIRELESS_SSID_COLUMNS))
```

File: nautobot_chatops_ipfabric/worker.py [hunk 8]
```diff
@@ -666,16 +660,16 @@ def get_bgp_neighbors(dispatcher, device=None, snapshot_id=None, state=None):
...
-                neighbor["srcInt"],
-                neighbor["localAddress"],
-                neighbor["vrf"],
-                neighbor["neiHostname"],
-                neighbor["neiAddress"],
-                neighbor["neiAs"],
-                neighbor["state"],
-                neighbor["totalReceivedPrefixes"],
+                neighbor[IpFabric.BGP_NEIGHBORS_COLUMNS[0]],
+                neighbor[IpFabric.BGP_NEIGHBORS_COLUMNS[1]],
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:664]:
```py
( neighbor[IpFabric.BGP_NEIGHBORS_COLUMNS[i]] for i in range(len(IpFabric.BGP_NEIGHBORS_COLUMNS))
```

File: nautobot_chatops_ipfabric/worker.py [hunk 9]
```diff
@@ -261,44 +293,53 @@ def interfaces(dispatcher, device=None, metric=None):
...
     )
 
     dispatcher.send_large_table(
         ["IntName", "IN bps", "OUT bps"],
         [
             (
-                interface["intName"],
-                interface["inBytes"],
-                interface["outBytes"],
+                interface[IpFabric.INTERFACE_LOAD_COLUMNS[0]],
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:340]:
```py
(  interface[IpFabric.INTERFACE_LOAD_COLUMNS[i]] for i in range(len(IpFabric.INTERFACE_LOAD_COLUMNS))
```

File: nautobot_chatops_ipfabric/worker.py [hunk 10]
```diff
@@ -309,29 +350,38 @@ def get_int_load(dispatcher, device, snapshot_id):
...
     )
 
     dispatcher.send_large_table(
         ["IntName", "Error %", "Error Rate"],
         [
             (
-                interface["intName"],
-                interface["errPktsPct"],
-                interface["errRate"],
+                interface[IpFabric.INTERFACE_ERRORS_COLUMNS[0]],
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:382]:
```py
(  interface[IpFabric. INTERFACE_ERRORS_COLUMNS[i]] for i in range(len(IpFabric. INTERFACE_ERRORS_COLUMNS))
```

File: nautobot_chatops_ipfabric/worker.py [hunk 11]
```diff
@@ -165,56 +176,68 @@ def get_snapshot(dispatcher):
...
-                device.get("hostname") or "(empty)",
-                device.get("siteName") or "(empty)",
-                device.get("vendor") or "(empty)",
-                device.get("platform") or "(empty)",
-                device.get("model") or "(empty)",
-                device.get("memoryUtilization") or "(empty)",
-                device.get("version") or "(empty)",
-                device.get("sn") or "(empty)",
-                device.get("loginIp") or "(empty)",
+                device.get(IpFabric.INVENTORY_COLUMNS[0]) or IpFabric.EMPTY,
```
Comment (chadell) [inline nautobot_chatops_ipfabric/worker.py:232]:
```py
(host.get(IpFabric. INVENTORY_COLUMNS[i], IpFabric.EMPTY) for i in range(len(IpFabric. INVENTORY_COLUMNS))
```

**Review by chadell — APPROVED** [review]
LGTM
added a cosmetic change suggestion

File: nautobot_chatops_ipfabric/worker.py (hunk 5 above)
Comment (pke11y) [reply nautobot_chatops_ipfabric/worker.py:863]: Nice one!

Comment (pke11y) [reply nautobot_chatops_ipfabric/worker.py]: Turns out there is no tuple comprehension. That syntax is reserved for generator objects. Just used list comprehension instead.

**Pull request merged**